text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,264 @@
+
+## Parameterizing SQL Commands
+
+SQL queries usually appear in descriptions of `db` spans. In a SQL query (or any SQL statement) a parameter is any value that is either generated randomly, or supplied from a variable in the code. The following symbols are *not* considered parameters, and should not be altered or re-ordered:
+
+- column names
+- table names
+- aliases
+
+### Simple Parameters
+
+For individual parameters, **the parameterizer must replace the parameter with a single unquoted `?` character**.
+
+Given:
+
+```sql
+SELECT "countries"."id", "countries.name"
+FROM "countries"
+WHERE "countries"."code" = 'CA';
+```
+
+Return:
+
+```sql
+SELECT "countries"."id", "countries.name"
+FROM "countries"
+WHERE "countries"."code" = ?;
+```
+
+### Parameter Lists (`IN` Clauses)
+
+For lists of parameters, the parameterizer must replace the entire list with a single unquoted `?` character in parentheses.
+
+Given:
+
+```sql
+SELECT "countries"."id", "countries.name"
+FROM "countries"
+WHERE "countries"."code" IN ("CA", "US");
+```
+
+Return:
+
+```sql
+SELECT "countries"."id", "countries.name"
+FROM "countries"
+WHERE "countries"."code" IN (?);
+```
+
+### `SAVEPOINT` Commands
+
+In `SAVEPOINT` and `ROLLBACK` commands, replace the savepoint's name with a single `?` character.
+
+Given:
+
+```sql
+ROLLBACK TO SAVEPOINT "s47890194282880_x50"'
+```
+
+Return:
+
+```sql
+ROLLBACK TO SAVEPOINT ?
+```
+
+## Parameterizing URLs
+
+A URL might appear in the name of a transaction, in the description of a span, as data in a breadcrumb, or some other situation. URLs might correspond to known web service endpoints, or arbitrary addresses. The parameterizer must use as much information as possible to parameterize URLs. Dynamic route segments (e.g., the `18` in `/users/18/info`), and dynamic hostname segments (e.g., the `us-east` in `us-east.service.com` vs. `ca-west` in `ca-west.service.com`) are considered parameters.
+
+### Known Route Segments
+
+A parameterizer that has knowledge of the route structure that the URL corresponds to must replace URL parameters with the correct route segment names.
+
+For example, the React Router SDK has knowledge of the URL structure of the app. Therefore, a parameterizer in the SDK should refer to the route structure, and supply the names of the URL segments. The same applies to web frameworks like Django. It is also in theory possible for a parameterizer to cross-reference a URL to a known service like GitHub against GitHub’s documentation, to determine whether outgoing requests match a specific API route.
+
+**The parameterizer must replace route parameters with the name of the parameter surrounded in curly braces.**
+
+Given:
+
+```
+GET http://someservice.com/users/tom/
+```
+
+If the parameterizer determines that `tom` is a route parameter named `:name`, it must return:
+
+```
+GET http://someservice.com/users/{name}
+```
```

Comment:
As with the section above trying to set a common substitution character, setting an arbitrary format may be confusing for users manually setting custom instrumentation. Is this something we're ok with?

Reply:
🤔 that's an interesting point! I think this might _reduce_ confusion, because if we publish a spec on how to parameterize strings, users with custom instrumentation can refer to our spec, and their parameterization will be compatible with what our system expects